sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,170 @@
+ * monotonically increasing fashion.
+ */
+public class ByteKeyRangeTracker extends RestrictionTracker<ByteKeyRange, ByteKey> {
+  private ByteKeyRange range;
+  @Nullable private ByteKey lastClaimedKey = null;
+  @Nullable private ByteKey lastAttemptedKey = null;
+
+  private ByteKeyRangeTracker(ByteKeyRange range) {
+    this.range = checkNotNull(range);
+  }
+
+  /**
+   * Instantiates a new {@link ByteKeyRangeTracker} with the specified range. The keys in the range
+   * are left padded to be the same length in bytes.
+   */
+  public static ByteKeyRangeTracker of(ByteKeyRange range) {
+    return new ByteKeyRangeTracker(ByteKeyRange.of(range.getStartKey(), range.getEndKey()));
+  }
+
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey, range);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    ByteKey rangeEndKey =
+        (nextKey.equals(range.getEndKey())) ? next(range.getEndKey(), range) : range.getEndKey();
+    ByteKeyRange res = ByteKeyRange.of(nextKey, rangeEndKey);
+    this.range = ByteKeyRange.of(range.getStartKey(), nextKey);
+    return res;
+  }
+
+  /**
+   * Attempts to claim the given key.
+   *
+   * <p>Must be larger than the last successfully claimed key.
+   *
+   * @return {@code true} if the key was successfully claimed, {@code false} if it is outside the
+   *     current {@link ByteKeyRange} of this tracker (in that case this operation is a no-op).
+   */
+  @Override
+  protected synchronized boolean tryClaimImpl(ByteKey key) {
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
+        key,
+        lastAttemptedKey);
+    checkArgument(
+        key.compareTo(range.getStartKey()) > -1,
+        "Trying to claim key %s before start of the range %s",
+        key,
+        range);
+    lastAttemptedKey = key;
+    // No respective checkArgument for i < range.to() - it's ok to try claiming keys beyond
+    if (!range.getEndKey().isEmpty() && key.compareTo(range.getEndKey()) > -1) {
+      return false;
+    }
+    lastClaimedKey = key;
+    return true;
+  }
+
+  /**
+   * Marks that there are no more keys to be claimed in the range.
+   *
+   * <p>E.g., a {@link DoFn} reading a file and claiming the key of each record in the file might
+   * call this if it hits EOF - even though the last attempted claim was before the end of the
+   * range, there are no more keys to claim.
+   */
+  public synchronized void markDone() {
+    lastAttemptedKey = range.getEndKey();
+  }
+
+  @Override
+  public synchronized void checkDone() throws IllegalStateException {
+    checkState(
```

Comment:
@jkff Notice here that I cannot presume the previous value of the endKey as Offset tracker does. Also because of lexicographic ordering there could be many valid values for this 'previous'.